debian/cdcat.1
.TH cdcat 1
.SH NAME
cdcat \- media catalog program
.SH SYNOPSIS
.B cdcat
.SH "DESCRIPTION"
This manual page documents briefly the
.BR cdcat ,
command.
This manual page was written for the Debian GNU/Linux distribution
because the original program does not have a manual page.
.PP
.B cdcat 
is a graphical, multiplatform media catalog program which scans the
directories/media you specify and makes a list of the filesystem (including
the tags of MP3s) and stores the result in a gzipped XML file.
.SH AUTHOR
This manual page was written by Jorge Bernal <koke@sindominio.net>,
for the Debian GNU/Linux system (but may be used by others).


ChangeLog
30. May 2014 Christoph Thielecke <crissi99@gmx.de>
	- proper initalization of iv
	- set catalog encryption to false on every DBCatalog constructor
	- change busy text for encryption

25. Jan 2014 Christoph Thielecke <crissi99@gmx.de>
	- typo on assiging cdrom path on unix fixed
	- stripping white spaces on date at whereisit xml import fixed

26. Oct 2013 Christoph Thielecke <crissi99@gmx.de>
	- csv import using pattern added (feature request #31)

29. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- russian translation generation fixed
	- version increased to 2.3.1

30. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- click on result entry does not select corresponding entry tree fixed
	- more keyboard shortcuts added
	- keyboard shortcut dialog added (help menu)

29. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- dont add empty exclude rule on rule list at selreadable dialog
	- make read blocksize configurable by define
	- update current node if no element (..) in content listview is selected fixed
	- delete keyboard shortcut added (del key)
	- display full path instead node name on delete confirmation dialog
	- make delete confirmation question better
	- display full path instead node name on size dialog

26. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- icons for encryption related buttons added
	- compare on TestInStream constructor fixed
	- missing initalization of vars (bTAGreaded, bTAGexist, tnumber) at ReadMp3Tag class on constructor added
	- reading old catalog without sortedBy attribute fails fixed
	- use qstring instead char * for what and err string at getStr2, getDouble2

25. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- window title on borrow dialog fixed
	- use initalization list for init local vars instead inside constructor on mediainfo dialog
	- resource leak on insertDB at out of memory error fixed
	- use initalization list for init local vars instead inside constructor on colorsettings dialog
	- typo on commwidget contructor on init a var fixed
	- missing initalization of etype var on constructors added
	- code cleanup at getFullPath() on Node class
	- missing initalization of information var at lineobject class on constructor added
	- assignment of fileinfo var on DBFile constructor fixed
	- missing initalization of vars (length, last, alternate) at MString class on constructor added
	- not nessary var errormsg removed
	- missing initalization of vars (usefileinfo, useExternalContentViewer, storeExifData, commentDockPos_x, commentDockPos_y, commentWindowPos_x, commentWindowPos_y, commentDockSize_width, commentDockSize_height, contentWindowPos_x, contentWindowPos_y, contentWindowSize_width, contentWindowSize_height commentWindowSize_width, commentWindowSize_height) at CdcatConfig class on constructor added
	- missing initalization of vars (founded, pww, error, errptr, dirname, filename, comment, find_category, tartist, ttitle, talbum, tcomment, content, allmedia,
allowner, dateStartChecked, dateEndChecked, sizeMinChecked, sizeMaxChecked, findInArchivesChecked, size_min, size_max) at seekEngine class on constructor added
	- typo at seeke on finddialog class fixed

23. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- regression on saving thumbnail data to catalog file fixed
	- dont add history item on save/save as if already in list
	- add media fails after previous cancelled scan fixed

22. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- crash on reading fileinfo in some cases fixed
	- crash on reading file content fixed
	- memleak at detecting cd/dvd type fixed
	- memleak on open cd device fixed
	- endless loop fixed on scanning avi info at file with broken avi header

14. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- memleak at dirview class fixed
	- memleaks at recent menu fixed
	- cleanup of mediainfo fixed

13. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- memleak on reading catalog at read content fixed
	- some debug info disabled
	- memleak at mediainfo fixed
	- memleak at config class fixed
	- memleak at HDirectoryView fixed
	- double free fixed
	- memleak on setting splitter style fixed
	- memleaks on exit fixed
	- memleaks on closing catalog fixed

09. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- initial russian translation added (thanks to ferm32 (Vitaly A. Kuzmin))

08. Aug 2013 Christoph Thielecke <crissi99@gmx.de>
	- displaying nfo pattern on debug info disabled
	- return type of isIdentical changed from int to bool
	- media name not changed after dialog on add media fixed (#53)
	- check for existing media name not only for cd/dvd added
	- missing initialisation of QTextCodec resulting in unability to run mediainfo on accentuated filenames fixed (#55)
	- missing close file handle on xml export fixed (#56)
	- wrong read mode of catalog on xml export fixed (#57)

12. Jan 2012 Christoph Thielecke <crissi99@gmx.de>
	- menu entries/context menu entries for enable/disable encryption added
	- password confirmation on set/change encryption password added

11. Jan 2012 Christoph Thielecke <crissi99@gmx.de>
	- display compiled in features in about dialog added
	- support for encrypted catalog added (requires libcrypto++, can be disabled on compile time)

03. Jan 2012 Christoph Thielecke <crissi99@gmx.de>
	- Q_OBJECT macro on ArchiveFile class added
	- cdcat homepage link at info dialog fixed
	- gpl page link at info dialog added
	- crash on cancel of add link dialog fixed (thanks to Alain Aupeix)
	- display path of selected search result entry in status bar added

02. Jan 2012 Christoph Thielecke <crissi99@gmx.de>
	- links now working in regex info dialog (thanks to Alain Aupeix)
	- wikipedia link fixed in regex info dialog

31. Dec 2012 Christoph Thielecke <crissi99@gmx.de>
	- sort catalog by name (descending) added
	- sort catalog by media type (ascending/descending) added
	- export single media to xml added
	- import of cdcat exported xml added (e.g. singe media)

30. Dec 2012 Christoph Thielecke <crissi99@gmx.de>

src/cdcat.cpp
/****************************************************************************
*                      Hyper's CD Catalog
*  A multiplatform qt and xml based catalog program
*
*  Author    : Peter Deak (hyperr@freemail.hu)
*  License   : GPL
*  Copyright : (C) 2003 Peter Deak
****************************************************************************/

#include "cdcat.h"

#include "dbase.h"
#include "mainwidget.h"
#include "icons.h"

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
// #include <iostream>

#include <QCommandLineOption>
#include <QCommandLineParser>
#include <QApplication>
#include <QTranslator>
#include <QFont>
#include <QMessageBox>
#include <QDir>
#include <QList>
#include <QFileInfo>
#include <QTextCodec>
#include <QDebug>
#include <QResource>

/* Own debug output {{{ */
bool *init_debug_info() {
    if (DEBUG_INFO_ENABLED == NULL) {
        DEBUG_INFO_ENABLED = new bool();
    }
    return DEBUG_INFO_ENABLED;
}

#ifndef QT_NO_DEBUG
void myMessageOutput(QtMsgType type, const QMessageLogContext &context, const QString &msg) {
    QByteArray localMsg = msg.toLocal8Bit();

    switch (type) {
    case QtDebugMsg:
        fprintf(stderr, "Debug: %s (%s:%u, %s)\n", localMsg.constData(), context.file, context.line, context.function);
        break;
    case QtWarningMsg:
        fprintf(stderr, "Warning: %s (%s:%u, %s)\n", localMsg.constData(), context.file, context.line, context.function);
        break;
    case QtCriticalMsg:
        fprintf(stderr, "Critical: %s (%s:%u, %s)\n", localMsg.constData(), context.file, context.line, context.function);
        break;
    case QtFatalMsg:
        fprintf(stderr, "Fatal: %s (%s:%u, %s)\n", localMsg.constData(), context.file, context.line, context.function);
        abort();
    }
}
#endif
/* }}} */

int main( int argi, char **argc ) {

#ifndef QT_NO_DEBUG
    #warning ====> installing own message handler
    qInstallMessageHandler( myMessageOutput );
#endif

    QTextCodec::setCodecForLocale( QTextCodec::codecForName( "UTF-8" ));

    QApplication app( argi, argc );
    // QCoreApplication::setOrganizationDomain();
    QCoreApplication::setOrganizationName(PROGRAM_NAME);
    QCoreApplication::setApplicationName(PROGRAM_NAME);
    QCoreApplication::setApplicationVersion(VERSION);

    /* Internationalization {{{ */
    QTranslator translator;

#ifdef QT_NO_DEBUG
    qDebug() << "translator.load, resource file:" << translator.load( "cdcat_" + QLocale::system().name(), ":/lang" );
#else

#if defined(_WIN32) || defined(Q_OS_MAC) || defined(_OS2)
    QString langpath( applicationDirPath( argc ) + "/lang/cdcat_" );
    langpath += cconfig->lang;
    langpath += ".qm";
#endif
    qDebug() << "translator.load, ./lang/:" << translator.load( "cdcat_" + QLocale::system().name(), "./lang" );
#endif
    app.installTranslator( &translator );
    /* }}} */

    /* Declare command line parameters {{{ */
    QCommandLineParser parser;
    parser.setApplicationDescription(QCoreApplication::translate("helpDialog", "The %1 is graphical (QT based) multiplatform (Linux/Windows) catalog program which scans the directories/drives you want and memorize the filesystem including the tags of mp3's and other data and store it in a small file. The database is stored in a gzipped XML format, so you can hack it, or use it if necessary :-).)" ).arg(PROGRAM_NAME));

    // parser.addHelpOption();
    // Can not be translated?
    /* print help {{{ */
    QCommandLineOption optionPrintHelp(QStringList() << "h" << "help",
        QCoreApplication::translate("main_cli", "Displays this help.")
    );
    Q_ASSERT(parser.addOption(optionPrintHelp));
    /* }}} */

    // parser.addVersionOption();
    /* print version {{{ */
    QCommandLineOption optionPrintVersion(QStringList() << "V" << "version",
        QCoreApplication::translate("main_cli", "Displays version information.")
    );
    Q_ASSERT(parser.addOption(optionPrintVersion));
    /* }}} */

    parser.addPositionalArgument("db_filename",
        QCoreApplication::translate("main_cli", "Optional database file, if given, it will be used as database.")
    );

    /* Import database into the database {{{ */
    QCommandLineOption optionImportFile(QStringList() << "i" << "import-file",
        QCoreApplication::translate("main_cli", "Import %1 in the format specified by the %2 parameter into the database.")
                .arg("<" + QCoreApplication::translate("seekEngine", "file") + ">")
                .arg("--import-format")
        ,
        QCoreApplication::translate("seekEngine", "file")
    );
    Q_ASSERT(parser.addOption(optionImportFile));

    QCommandLineOption optionImportFormat(QStringList() << "f" << "import-format",
        QCoreApplication::translate("main_cli", "Format used by the %1 parameter."
            " Use the %2 parameter to get a list of all supported import formats."
            " Default is %3.")
                .arg("--import-file")
                .arg("--import-list")
                .arg("\"Whereisit\"")
        ,
        QCoreApplication::translate("main_cli", "format"),
        "Whereisit"
    );
    Q_ASSERT(parser.addOption(optionImportFormat));

    QCommandLineOption optionImportList(QStringList() << "l" << "import-list",
        QCoreApplication::translate("main_cli", "Print a list of all supported import formats and exit.")
    );
    Q_ASSERT(parser.addOption(optionImportList));
    /* }}} */

    /* Export database to other format {{{ */
    QCommandLineOption optionExportFile(QStringList() << "e" << "export-file",
        QCoreApplication::translate("main_cli", "Export the contents of the database to %1 in the format specified by the %2 parameter.")
            .arg("<" + QCoreApplication::translate("seekEngine", "file") + ">")
            .arg("--export-format")
        ,
        QCoreApplication::translate("seekEngine", "file")
    );
    Q_ASSERT(parser.addOption(optionExportFile));

    QCommandLineOption optionExportFormat(QStringList() << "t" << "export-format",
        QCoreApplication::translate("main_cli", "Format used by the %1 parameter."
            " Use the %2 parameter to get a list of all supported export formats."
            " Default is %3.")
                .arg("--export-file")
                .arg("--export-list")
                .arg("\"Whereisit\"")
        ,
        QCoreApplication::translate("main_cli", "format"),
        "Whereisit"
    );
    Q_ASSERT(parser.addOption(optionExportFormat));
    /* }}} */

    QCommandLineOption optionConfigFile(QStringList() << "c" << "config-file",
        QCoreApplication::translate("main_cli", "Configuration file to use."),
        QCoreApplication::translate("main_cli", "file")
    );
    // Q_ASSERT(parser.addOption(optionConfigFile));

    QCommandLineOption optionSwitchBatchMode(QStringList() << "b" << "batch",
        QCoreApplication::translate("main_cli", "Run in non iterative batch mode. This mode is CLI only which means no GUI is going to appear.")
    );
    Q_ASSERT(parser.addOption(optionSwitchBatchMode));

    QCommandLineOption optionExportList(QStringList() << "L" << "export-list",
        QCoreApplication::translate("main_cli", "Print a list of all supported export formats and exit.")
    );
    Q_ASSERT(parser.addOption(optionExportList));
    /* }}} */

    /* Evaluate command line parameters {{{ */
    parser.process(app);

    const QStringList args = parser.positionalArguments();

    if (parser.isSet(optionPrintHelp)) {
        parser.showHelp(0);
    }

    if (parser.isSet(optionPrintVersion)) {
        // parser.showVersion();
        // FIXME: This function was introduced in Qt 5.4.
        fprintf(stdout, "%s %s\n",
            qPrintable(QCoreApplication::applicationName()),
            qPrintable(QCoreApplication::applicationVersion())
        );
        return 0;
    }

    /* Evaluate command line parameters: Error checking {{{ */
    if (args.size() > 1) {
        fprintf(stderr, "%s\n", qPrintable(
            QCoreApplication::translate("main_cli", "Error: You can only specify one %1.")
                .arg("\"db_filename\"" + args.size() )
            )
        );
        parser.showHelp(1);
    }
    /* }}} */

    /* }}} */

    CdCatConfig *cconfig;

    // if (parser.isSet(optionConfigFile)) {
        // cconfig = new CdCatConfig(parser.value(optionConfigFile));
    // } else {
        cconfig = new CdCatConfig(NULL);
    // }

    if (args.size() == 1) {
        QByteArray filename_a = args.at(0).toUtf8();
        char* filename = filename_a.data();
        qDebug() << "filename:" << filename;
        cconfig->setParameter( filename );
    }

    cconfig->readConfig();

    // FIXME remove
    DEBUG_INFO_ENABLED = init_debug_info();
    *DEBUG_INFO_ENABLED = cconfig->debug_info_enabled;
    qDebug() << "DEBUG_INFO_ENABLED:" << QString(*DEBUG_INFO_ENABLED ? "true" : "false");

    int ret_val = 2;
    if (parser.isSet(optionSwitchBatchMode)) {
    } else if (parser.isSet(optionImportList)) {
    } else if (parser.isSet(optionExportList)) {
    } else {
        /* Create GUI {{{ */
        init_icon_base();

        CdCatMainWidget mw( cconfig, &app, 0, "MainWindow" );

        cconfig->defaultfont = app.font();
        if (cconfig->ownfont) {
            QFont font;
            font.setPointSize( cconfig->fsize );
            app.setFont( font );
        }

        mw.show();

        ret_val = app.exec();
        deinit_icon_base();
        /* }}} */
    }
    delete cconfig;

    return ret_val;
}

char *mstr( const char *imp ) {
    char *x;

    if (imp == NULL) {
        return NULL;
    }
    x = new char[strlen( imp ) + 1];
    strcpy( x, imp );
    return x;
}

QString applicationFilePath( char **argc ) {
#if defined(_WIN32) || defined(_OS2)
    return QDir::cleanPath( QFile::decodeName( "cdcat" ));
#else

    QString absPath;
    QString tmp( argc[0] );
    if (tmp.startsWith( "/" )) {
        /*
         * If argc[0] starts with a slash, it is already an absolute
         * file path.
         */
        absPath = tmp;
    } else if (tmp.indexOf( '/' ) != -1) {
        /*
         * If argc[0] contains one or more slashes, it is a file path
         * relative to the current directory.
         */
        absPath = QDir::current().absoluteFilePath( tmp );
    } else {
        /*
         * Otherwise, the file path has to be determined using the
         * PATH environment variable.
         */
        char *pEnv = getenv( "PATH" );
        QStringList paths = QString( pEnv ).split( QChar( ':' ));
        QStringList::const_iterator p = paths.begin();
        while (p != paths.end()) {
            QString candidate = QDir::current().absoluteFilePath( *p + "/" + tmp );
            if (QFile::exists( candidate )) {
                absPath = candidate;
                break;
            }
            ++p;
        }
    }

    absPath = QDir::cleanPath( absPath );
    if (QFile::exists( absPath )) {
        //		return resolveSymlinks( absPath );
        return absPath;
    } else {
        return QString::null;
    }
#endif
}

QString applicationDirPath( char **argc ) {
    return QFileInfo( applicationFilePath( argc )).path();
}
